### PR TITLE
Check restrictions on how `this` values can be used in constructors.

### PR DESCRIPTION
**Pull request by sjrd**
Essentially, `this` cannot be used before the super constructor call, except to assign fields (but not reading).

These restrictions match those enforced by the JVM for bytecode.

While we never intended `this` values to be restricted in this way, in practice we never produced IR that would violate those restrictions.

In the future, we may leverage those restrictions to validate strict field initialization. If a field *needs* to be initialized before it is ever read (for example, because it has a non-nullable reference type), then we can do so: we only have to check that it is definitely initialized before the super constructor call. The restrictions implemented in this commit ensure that nobody can read the field before the super constructor call starts.

### CODE REVIEW

**Comment by sjrd** [comment]
@gzm0 So I could indeed fix the `liftedTree` problem by making lifted methods `static` when possible. This is mostly the same analysis that the JVM backend does in `Delambdafy`, although it is simpler. They make differences between `isLiftedMethod` methods and `isDelambdafyTarget` methods, for some reason.

It's still quite problematic because it does not retroactively fix code that was compiled with previous versions of the compiler. I don't know what to do about that. It's super rare; one almost has to write contrived code to run into the issue. But it's still IR that was considered valid before, that may be published somewhere in some library, and that we would "abandon" as being invalid now.

**Comment by gzm0** [comment]
> But it's still IR that was considered valid before, that may be published somewhere in some library, and that we would "abandon" as being invalid now.

Do we have an understanding of whether such IR actually works? (Especially in contrived downstream usage scenarios).

Or in other words: Can we construct an end-to-end example of usage of a library that has such IR causes some form of user visible problem (linker or code crashing)?

If yes, we have a good case to declare this a bug and essentially say the more restrictive class def checking catches more bugs in libraries (even if they were caused by Scala.js itself).
If no, the reason why not hopefully gives us more insight in what specifically makes this OK (unfortunately, no guarantee though, that it gives us a path forward :shrug:).

**Comment by gzm0** [comment]
As an alternative, we could investigate options to turn the restrictions down. For example, we could apply the restrictions only if there exists a non-nullable field in the ancestry chain.

I guess the issues with many of these is going to be:
- They rely on infos of the ancestry chain.
- Adding a non-nullable field to a class is going to be a binary incompatible change, making it essentially impossible to do it as an optimization in the compiler.

**Comment by sjrd** [comment]
> Do we have an understanding of whether such IR actually works? (Especially in contrived downstream usage scenarios).
> 
> Or in other words: Can we construct an end-to-end example of usage of a library that has such IR causes some form of user visible problem (linker or code crashing)?

I don't see why such IR wouldn't work. I always thought this to be valid. We never made the assumption that `this` was not fully usable before the super constructor. Here I am definitely proposing to make invalid something that was considered valid before (not just that we did not check it). I'm only thinking about this because I saw that it was a requirement on the JVM, and that safety of non-defaultable fields must rely on that property.

> Adding a non-nullable field to a class is going to be a binary incompatible change, making it essentially impossible to do it as an optimization in the compiler.

That's probably reason enough not to design the system this way, IMO.

**Comment by gzm0** [comment]
> That's probably reason enough not to design the system this way, IMO.

I initially was going to agree with this, but maybe there can still be value. For example, changing to non-nullable would be a binary compatible change for final classes (granted, final is not in the IR). So for example, data classes could profit from this.

Two more ideas I had:

- Investigate different means of representing field initialization / construction (e.g. allow an `rhs` to field defs, but in which context that would be evaluated, who knows).
- Introduce the restrictions only for the next version of IR and deal with it in the entire linker pipeline (e.g. we could widen non-nullable fields to nullable and introduce runtime checks at usage sites).

Both of these seem rather hard to do if feasible at all. Otherwise I think our options boil down to:

- Don't do this (abandon PR)
- Break backwards compat (complete and merge PR)
- Intricate scheme with flags and opt-in to ensure backwards compat.
- SJS v2

**Comment by sjrd** [comment]
I ran an "audit" of all the .sjsir files found in all the published Scala.js libraries (on Maven Central). I used Scaladex to get a list of all the artifacts with their latest version. For each, I used the Coursier API to fetch their jar, then I ran the `ClassDefChecker` of this PR on all the `.sjsir` files in them.

There were 7029 artifacts, and none of them yielded a "restricted `this`" error.

Script and results here:
https://gist.github.com/sjrd/6bafd71bbc1fe8f99ee80271c8a42553

IMO that should be enough evidence that we can just abandon IR that has restricted `this`. WDYT?

---

Ironically, there were 3 errors about the constructor call discipline in Akka.js. But that's in IR that they manually patch. There were also two errors in synthetic magic traits of the Scala 3 library. Those are .sjsir files that we never actually load, AFAICT.

**Comment by gzm0** [comment]
> IMO that should be enough evidence that we can just abandon IR that has restricted this. WDYT?

Agreed.

**Comment by sjrd** [comment]
All right. I switched the order of the commits, and cleaned up comments and commit messages. Ready for an actual review now.

**Comment by ekrich** [comment]
The audit is very nice as it pulled all the Scala.js 1.x artifacts including all the older versions even published for Scala 2.11.

File: compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala [hunk 1]
```diff
@@ -248,6 +248,138 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
...
+           * but we do need to make note of the lifted body method in case it refers to `this`.
+           */
+          if (currentMethod.exists)
+            addReference(from = currentMethod, to = target.symbol)
+        case Apply(sel @ Select(This(_), _), args) if isStatifyCandidate(sel.symbol) =>
+          if (currentMethod.exists)
+            addReference(from = currentMethod, to = sel.symbol)
+          super.traverseTrees(args)
+        case This(_) =>
+          if (currentMethod.exists && tree.symbol == currentMethod.enclClass)
```
Comment (gzm0) [inline compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:376]: What does it mean if `tree.symbol != currentMethod.enclClass`?

File: compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala [hunk 2]
```diff
@@ -248,6 +248,138 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
...
+          // mark `symbol` as referring to `this`
+          if (result.add(symbol)) {
+            // `symbol` was not yet in the set; propagate further
+            for {
+              referrers <- methodReferences.get(symbol)
+              referrer <- referrers
+            } {
+              // referrer calls `symbol`, so it transitively references `this`
+              rec(referrer)
+            }
```
Comment (gzm0) [inline compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:344]:
```suggestion
            methodReferences.getOrElse(symbol, Nil).foreach(rec(_))
```

For readability? (IMO the nested comment is not necessary).

File: linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala [hunk 3]
```diff
@@ -915,6 +932,8 @@ private final class ClassDefChecker(classDef: ClassDef,
           reportError(i"Cannot find `this` in scope")
         else if (tree.tpe != env.thisType)
           reportError(i"`this` of type ${env.thisType} typed as ${tree.tpe}")
+        else if (env.isThisRestricted)
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:935]: Why the else? Wouldn't we want to report errors exhaustively (if sensible)?

File: linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala [hunk 4]
```diff
@@ -704,10 +719,12 @@ private final class ClassDefChecker(classDef: ClassDef,
       case StoreModule() =>
         if (!classDef.kind.hasModuleAccessor)
           reportError(i"Illegal StoreModule inside class of kind ${classDef.kind}")
-        if (!env.inConstructor)
+        else if (!env.inConstructor)
           reportError(i"Illegal StoreModule outside of constructor")
-        if (env.thisType == NoType) // can happen before JSSuperConstructorCall in JSModuleClass
+        else if (env.thisType == NoType) // can happen before JSSuperConstructorCall in JSModuleClass
           reportError(i"Cannot find `this` in scope for StoreModule()")
+        else if (env.isThisRestricted)
```
Comment (gzm0) [inline linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:726]: Why the change to else? Wouldn't we want to report errors exhaustively (if sensible)?

**Review by gzm0 — CHANGES_REQUESTED** [review]

File: compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala (hunk 1 above)
Comment (sjrd) [reply compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:371]: It can apparently be a static module to be loaded with `genLoadModule`. I added a comment.

File: linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala (hunk 4 above)
Comment (sjrd) [reply linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala]: I figured some of these errors might lead to other errors. But I reverted to not having `else`s. Same below.

**Review by sjrd — COMMENTED** [review]
Updated.

**Review by gzm0 — APPROVED** [review]
